File: hw8/client/src/app/models/api-status.ts
enum ApiStatusState {
    INITIAL,
    LOADING,
    SUCCESS,
    ERROR
}

export class ApiStatus {
    private state = ApiStatusState.INITIAL;

    errorMessage: string = null;

    private isState(state: ApiStatusState): boolean {
        return this.state === state;
    }

    isInitial(): boolean {
        return this.isState(ApiStatusState.INITIAL);
    }

    isLoading(): boolean {
        return this.isState(ApiStatusState.LOADING);
    }

    isSuccess(): boolean {
        return this.isState(ApiStatusState.SUCCESS);
    }

    isError(): boolean {
        return this.isState(ApiStatusState.ERROR);
    }

    isCompleted(): boolean {
        return this.isError() || this.isSuccess();
    }

    reset(): void {
        this.state = ApiStatusState.INITIAL;
        this.errorMessage = null;
    }

    loading(): void {
        this.state = ApiStatusState.LOADING;
        this.errorMessage = null;
    }

    success(): void {
        this.state = ApiStatusState.SUCCESS;
        this.errorMessage = null;
    }

    error(errorMessage: string): void {
        this.state = ApiStatusState.ERROR;
        this.errorMessage = errorMessage;
    }

    getError(): string {
        return this.errorMessage;
    }
}

File: hw9/client/StockTradingApp/app/src/main/java/com/rochakgupta/stocktrading/common/api/ApiStatus.java
package com.rochakgupta.stocktrading.common.api;

public class ApiStatus {
    private enum Status {
        INITIAL,
        LOADING,
        SUCCESS,
        ERROR
    }

    private Status status;

    public ApiStatus() {
        status = Status.INITIAL;
    }

    public void loading() {
        status = Status.LOADING;
    }

    public void success() {
        status = Status.SUCCESS;
    }

    public void error() {
        status = Status.ERROR;
    }

    public boolean isInitial() {
        return status == Status.INITIAL;
    }

    public boolean isLoading() {
        return status == Status.LOADING;
    }

    public boolean isSuccess() {
        return status == Status.SUCCESS;
    }

    public boolean isError() {
        return status == Status.ERROR;
    }

    public boolean isComplete() {
        return isSuccess() || isError();
    }
}
